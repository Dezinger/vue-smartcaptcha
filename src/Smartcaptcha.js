import smartcaptcha from './smartcaptcha-wrapper'

export default {
  name: 'VueSmartcaptcha',
  props: {
    sitekey: {
      type: String,
      required: true
    },
    hl: {
      type: String,
      default: ''
    },
    test: {
      type: Boolean,
      default: false
    },
    webview: {
      type: Boolean,
      default: false
    },
    invisible: {
      type: Boolean,
      default: false
    },
    shieldPosition: {
      type: String
    },
    hideShield: {
      type: Boolean
    },
    loadSmartcaptchaScript: {
      type: Boolean,
      default: false
    },
    smartcaptchaScriptId: {
      type: String,
      default: '__SMARTCAPTCHA_SCRIPT'
    },
    smartcaptchaHost: {
      type: String,
      default: 'smartcaptcha.yandexcloud.net'
    }
  },
  beforeMount () {
    if (this.loadSmartcaptchaScript) {
      if (!document.getElementById(this.smartcaptchaScriptId)) {
        // Note: vueSmartcaptchaApiLoaded load callback name is per the latest documentation
        const script = document.createElement('script')
        script.id = this.smartcaptchaScriptId
        script.src = `https://${this.smartcaptchaHost}/captcha.js?render=onload&onload=vueSmartcaptchaApiLoaded`
        script.async = true
        script.defer = true

        document.head.appendChild(script)
      }
    }
  },
  mounted () {
    smartcaptcha.checkSmartcaptchaLoad()
    const opts = {
      ...this.$props,
      callback: this.emitVerify// ,
      // 'expired-callback': this.emitExpired,
      // 'error-callback': this.emitError
    }
    const container = this.$slots.default ? this.$el.children[0] : this.$el
    smartcaptcha.render(container, opts, id => {
      this.$widgetId = id
      this.$emit('render', id)
    })
  },
  methods: {
    reset () {
      smartcaptcha.reset(this.$widgetId)
    },
    execute () {
      smartcaptcha.execute(this.$widgetId)
    },
    emitVerify (response) {
      this.$emit('verify', response)
    }// ,
    /*
    emitExpired () {
      this.$emit('expired')
    },
    emitError () {
      this.$emit('error')
    }
    */
  },
  render (h) {
    return h('div', {}, this.$slots.default)
  }
}
