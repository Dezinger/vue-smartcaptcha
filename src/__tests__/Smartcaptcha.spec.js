import smartcaptcha, { WIDGET_ID } from '../smartcaptcha-wrapper'

import Smartcaptcha from '../Smartcaptcha'
import { mount } from '@vue/test-utils'

jest.mock('../smartcaptcha-wrapper')

const SITE_KEY = 'sitekey'
const createWrapper = propsData => {
  return mount(Smartcaptcha, { propsData })
}

describe('Smartcaptcha', () => {
  const wrapper = createWrapper({ sitekey: SITE_KEY })

  it('Should render SmartCAPTCHA', () => {
    expect(smartcaptcha.checkSmartcaptchaLoad).toBeCalled()
    expect(smartcaptcha.render.mock.calls[0][0]).toBe(wrapper.vm.$el)
    expect(smartcaptcha.render.mock.calls[0][1]).toMatchSnapshot('SmartCAPTCHA options')
  })

  /*
  it('Emit events', () => {
    expect(wrapper.emitted()).not.toContainKey('verify')
    wrapper.vm.emitVerify()
    expect(wrapper.emitted().verify).toBeTruthy()

    expect(wrapper.emitted()).not.toContainKey('expired')
    wrapper.vm.emitExpired()
    expect(wrapper.emitted().expired).toBeTruthy()
  })
  */
  it('Can reset/execute', () => {
    expect(smartcaptcha.reset).not.toBeCalled()
    wrapper.vm.reset()
    expect(smartcaptcha.reset).toBeCalledWith(WIDGET_ID)

    expect(smartcaptcha.execute).not.toBeCalled()
    wrapper.vm.execute()
    expect(smartcaptcha.execute).toBeCalledWith(WIDGET_ID)
  })

  it('will load smartcaptcha script when `loadSmartcaptchaScript` set to `true`', () => {
    const id = 'SMARTCAPTCHA_SCRIPT_ID'
    createWrapper({
      sitekey: SITE_KEY,
      loadSmartcaptchaScript: true,
      smartcaptchaScriptId: id
    })
    expect(document.getElementById(id)).not.toBe(null)
  })
})
