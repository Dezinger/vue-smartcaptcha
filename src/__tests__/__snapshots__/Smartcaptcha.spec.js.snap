// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smartcaptcha Should render SmartCAPTCHA: SmartCAPTCHA options 1`] = `
Object {
  "callback": [Function],
  "hideShield": false,
  "hl": "",
  "invisible": false,
  "loadSmartcaptchaScript": false,
  "shieldPosition": undefined,
  "sitekey": "sitekey",
  "smartcaptchaHost": "smartcaptcha.yandexcloud.net",
  "smartcaptchaScriptId": "__SMARTCAPTCHA_SCRIPT",
  "test": false,
  "webview": false,
}
`;
